### PR TITLE
Replace the in-tree libc with the external copy

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit replaces the in-tree liblibc with the [external clone](https://github.com/rust-lang-nursery/libc) which has no evolved beyond the in-tree version in light of its [recent redesign](https://github.com/rust-lang/rfcs/pull/1291).

The primary changes here are:
- `src/liblibc/lib.rs` was deleted
- `src/liblibc` is now a submodule pointing at the external repository
- `src/libstd/sys/unix/{c.rs,sync.rs}` were both deleted having all bindings folded into the external liblibc.
- Many ad-hoc `extern` blocks in the standard library were removed in favor of bindings now being in the external liblibc.
- Many functions/types were added to `src/libstd/sys/windows/c.rs`, and the scattered definitions throughout the standard library were consolidated here.

At the API level this commit is **not a breaking change**, although it is only very lightly tested on the *BSD variants and is probably going to break almost all of their builds! Follow-up commits to liblibc should in theory be all that's necessary to get the build working on the *BSDs again.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @dhuseby, @semarie, @mneumann -- this is likely to break BSD builds, but if you want to test out I'd be more than willing to fix things up before it lands!

cc @rust-lang/libs -- fallout from the libc transition, especially on the Windows side of thing

r? @brson

**Comment by brson** [comment]
r=me

**Comment by brson** [comment]
Tagging this with relnotes to remind myself to mention the libc rewrite in the notes.

**Comment by semarie** [comment]
@alexcrichton thanks to cc me

yes, it seems to break openbsd, but it seems related to new-libc code. I will take a look.

**Comment by semarie** [comment]
I have made a PR for make openbsd (and bitrig, I hope) be able to build libc https://github.com/rust-lang-nursery/libc/pull/39

But I still have a weird error while trying to build rustc with my PR (stop in stage1 libstd building):
```
../src/libstd/lib.rs:268:12: 268:25 error: unused or unknown feature, #[deny(unused_features)] on by default
../src/libstd/lib.rs:268 #![feature(drop_in_place)]
```
whereas a version with "plain" liblibc build fine.

File: src/libstd/lib.rs [hunk 1]
```diff
@@ -214,10 +214,9 @@
        test(no_crate_inject, attr(deny(warnings))),
        test(attr(allow(dead_code, deprecated, unused_variables, unused_mut))))]
 
-// SNAP 1af31d4
-#![allow(unused_features)]
-// SNAP 1af31d4
-#![allow(unused_attributes)]
+#![cfg_attr(stage0, allow(unused_features))]
+#![cfg_attr(stage0, allow(unused_attributes))]
+#![cfg_attr(stage0, allow(improper_ctypes))]
```
Comment (semarie) [inline src/libstd/lib.rs]: @alexcrichton why this ? it breaks at least the openbsd build.

`src/libstd/lib.rs` has `drop_in_place` feature, but it used only on some targets. So removing `unused_attributes` will break stage1/stage2 for others.

Comment (alexcrichton) [inline src/libstd/lib.rs]: Hm, I'll look into fixing, I'm not a huge fan of allowing lints like this which are intentionally here to help clean up some code!

**Comment by alexcrichton** [comment]
@bors: p=1

The bots aren't cleaning themselves well, gonna be real painful to deal with this unless this lands first :\

**Comment by alexcrichton** [comment]
@bors: r=brson 07da457

here goes nothing

**Comment by dhuseby** [comment]
I've been in London at MozFest.  I'm trying to catch up now.

**Comment by mneumann** [comment]
@dhuseby : One thing you have to fix for FreeBSD is to define SIGSTKSZ. I did it for DragonFly (https://github.com/rust-lang-nursery/libc/pull/60), but didn't know the exact value for FreeBSD.

**Comment by dhuseby** [comment]
@mneumann thanks, i believe it's fixed now.

**Pull request merged**